### PR TITLE
build: ESLint rule to check/add license headers

### LINKED ISSUE #11701: [Feature]: Create an EsLint rule to check/add Licence headers
Currently when new files are added we need to manually review, verify and add the correct licence header.
Having a rule to automatically do this will save some development time/friction. 
Example of the header that should exits on each file (.ts,.mjs,.cjs)
```js
/**
 * @license
 * Copyright 2018 Google Inc.
 * SPDX-License-Identifier: Apache-2.0
 */
 ```
If the header does not exist a new one should be added with the current year included.

Note: there are some files we would like to ignore as they are adapted from other open source projects and should keep their respective headers.

### PR DESCRIPTION
**Pull request by richestrinh**
**What kind of change does this PR introduce?**

An EsLint rule to check/add Licence headers

**Did you add tests for your changes?**

I wrote the test for this rule, but I was unable to locate the tests for the other rules (extension.js and prettier-comments.js).

Resolves #11701

This rule checks if a license header exists and fixes the 2 cases:
1. No license.
FIX: Adds license header.
2. Header comment that is not the license. 
FIX: Adds licensing as header.

**Other information**
There is another case where the license comment may already exist in the file, but not as a header. 
Shall I cover this case as well, and implement a fix to move that comment so it is a header?

Please let me know if anything else is needed! :)

### CODE REVIEW

**Comment by OrKoN** [comment]
cc @Lightning00Blade

File: tools/eslint/src/check-license.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+/**
+ * @license
+ * Copyright 2024 Google Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+const licenseHeader = `/**
+ * @license
+ * Copyright 2024 Google Inc.
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.js]: Let's use the current year and not hard-code 2024

File: .eslintrc.js [hunk 2]
```diff
@@ -149,6 +149,8 @@ module.exports = {
     'rulesdir/prettier-comments': 'error',
     // Enforces consistent file extension
     'rulesdir/extensions': 'error',
+    // Enforces license headers on files (.ts,.mjs,.cjs)
```
Comment (Lightning00Blade) [inline .eslintrc.js]: The files affected are controlled by EsLint so we can skip them.
```suggestion
    // Enforces license headers on files
```

File: tools/eslint/src/check-license.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
+/**
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.js]: Let's use TypeScript like `extensions.ts` or `use-using.ts`, to get better typing.

File: tools/eslint/src/check-license.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+          });
+        }
+        // Header exists, but is not license.
+        if (header && !header.value.includes('@license')) {
+          context.report({
+            node: node,
+            messageId: 'licenseRule',
+            fix(fixer) {
+              return fixer.insertTextBeforeRange(
+                [header.range[0], header.range[1]],
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.js]: We should be able to pass in `header.range` directly as it's a tuple of numbers `range?: [number, number];` 
But refer to my previous comment.

File: tools/eslint/src/check-license.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    return {
+      Program(node) {
+        if (header?.value.includes('@license')) {
+          return;
+        }
+
+        // Add header license
+        if (!header) {
+          context.report({
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.js]: I think we can use this for both case of `No Header` and `Header without @license`, as most likely we will still update/review it manually.

File: tools/eslint/src/check-license.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+ * Copyright 2024 Google Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */\n\n`;
+
+/**
+ * @type import("eslint").Rule.RuleModule
+ */
+
+const enforceLicenseRule = {
+  // module.exports = {
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.js]:
```suggestion
```

File: tools/eslint/src/check-license.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    fixable: 'code',
+    schema: [],
+    messages: {
+      licenseRule: 'Add license header.',
+    },
+  },
+  create(context) {
+    const sourceCode = context.sourceCode;
+    const comments = sourceCode.getAllComments();
+    const header = comments[0];
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.js]: Currently we take the first comment, which is OK, 
but we should check if the comment is of type `block` and if it's at the start of the file.
If one of the two is not `true` we should treat is as no header comment.

File: tools/eslint/src/check-license.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+    return {
+      Program(node) {
+        if (header?.value.includes('@license')) {
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.js]: We should check to include `@license`, `SPDX-License-Identifier: Apache-2.0` and
this RegEx `Copyright ([0-9]{4}) Google Inc.`. as we want a valid header as well.

**Review by Lightning00Blade — COMMENTED** [review]
@richestrinh Thank you for your contribution, overall looks good.
Please recreate the rule using TypeScript. (`prettier-comments.js` is old rule that we never migrated).
Also we currently have not set-up tests for EsLint rules so no need to provide any.
Everything else is in comment.

File: tools/eslint/src/check-license.js (hunk 5 above)
Comment (richestrinh) [reply tools/eslint/src/check-license.js]: I merged the cases together. 
The two cases had conflicting fixes, and if I understood your comment correct I have omitted the fixes as well.

File: tools/eslint/src/check-license.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        return false;
+      }
+    }
+
+    return {
+      Program(node) {
+        if (
+          header?.value.includes('@license') &&
+          header?.value.includes('SPDX-License-Identifier: Apache-2.0') &&
+          copyrightPattern.test(header?.value)
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.ts]:
```suggestion
          header &&
          header.value.includes('@license') &&
          header.value.includes('SPDX-License-Identifier: Apache-2.0') &&
          copyrightPattern.test(header.value)
```

File: tools/eslint/src/check-license.js (hunk 5 above)
Comment (Lightning00Blade) [reply tools/eslint/src/check-license.js]: Sorry about this, I though we can have the same logic for both, it seems like EsLint does not like it when you do that so we should split them as you have suggested.

File: tools/eslint/src/check-license.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  defaultOptions: [],
+  create(context) {
+    const sourceCode = context.sourceCode;
+    const comments = sourceCode.getAllComments();
+    const header =
+      comments[0]?.type === 'Block' && isHeaderComment(comments[0])
+        ? comments[0]
+        : null;
+
+    function isHeaderComment(comment: any) {
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.ts]:
```suggestion
    function isHeaderComment(comment: TSESTree.Comment) {
```

File: tools/eslint/src/check-license.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+      if (comment?.range[0] >= 0 && comment?.range[1] <= 88) {
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.ts]:
```suggestion
      if (comment && comment.range[0] >= 0 && comment.range[1] <= 88) {
```

File: tools/eslint/src/check-license.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
+/**
+ * @license
+ * Copyright 2024 Google Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {ESLintUtils} from '@typescript-eslint/utils';
+
+const createRule = ESLintUtils.RuleCreator(name => {
+  return `https://github.com/puppeteer/puppeteer/tree/main/tools/eslint/${name}.js`;
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.ts]: I think other places are wrong.
```suggestion
  return `https://github.com/puppeteer/puppeteer/tree/main/tools/eslint/${name}.ts`;
```

File: tools/eslint/src/check-license.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+});
+
+const copyrightPattern = /Copyright ([0-9]{4}) Google Inc\./;
+
+const enforceLicenseRule = createRule<[], 'licenseRule'>({
+  name: 'license',
```
Comment (Lightning00Blade) [inline tools/eslint/src/check-license.ts]: Name should match filename.
```suggestion
  name: 'check-license',
```

**Review by Lightning00Blade — COMMENTED** [review]
Please readd the fixer back (my mistake there) and remove the js file.
Other comments are small improvemnts.

**Comment by OrKoN** [comment]
Thanks @richestrinh for contributing this. To expand on @Lightning00Blade comment about tests, if you already wrote a test  for the rule, let's put it next to the rule file if that works (i.e., `<ruleName>.test.ts`  and let's add a command to run the test using the https://nodejs.org/api/test.html#running-tests-from-the-command-line 

I have triggered a lint run and it looks like there is one special case when the shebang comes before the license (as it should be). There could be other exception so it might be great to include that in the test.

File: .eslintrc.js [hunk 14]
```diff
@@ -149,6 +149,8 @@ module.exports = {
     'rulesdir/prettier-comments': 'error',
     // Enforces consistent file extension
     'rulesdir/extensions': 'error',
+    // Enforces license headers on files
+    'rulesdir/check-license': 'error',
```
Comment (OrKoN) [inline .eslintrc.js]: I think the current version is already useful so if you want @richestrinh we could flip this to "warn" and land this as the first version

**Comment by OrKoN** [comment]
I have prefixed the PR with `build: ` so that it does not show up in the changelog since it is not relevant for the end users.

File: tools/eslint/src/check-license.js (hunk 5 above)
Comment (richestrinh) [reply tools/eslint/src/check-license.js]: Sorry about this as well, I believe my updated fixer accounts for both cases as you mentioned

**Review by Lightning00Blade — APPROVED** [review]

**Comment by Lightning00Blade** [comment]
@richestrinh I am happy to merge as is, and iterate on changes/features. So please rebase and open for review.

**Pull request merged**